### PR TITLE
Add ThemedPalette

### PR DESCRIPTION
**Pull request by klnusbaum**
For #1798

This adds initial support for Theme based palettes. There's some things I'd like to discuss.

1. I'm pretty sure there's a way to avoid the `ThemeJSON` struct all together. Can someone recommend a good way to have Serde just automatically call the string -> Color conversion code I've written?
2. I'm not sure about the best way to organize the code. I've created a new theme module, but that may not really be worth it. We might just wanna lump it in with the Palette module. Open to discussion on this.
3. Open to suggestions on how we actually map colors to their corresponding `FlatShape`'s. I've given more fine grained control here than we have with the default formatter.

### CODE REVIEW

**Comment by klnusbaum** [comment]
I think things are looking a lot better right now. In an ideal world, I wouldn't have to annotate each `Color` field. Unfortunately, my understanding is that since the `Color` type and the `Deserialize` trait are both external crates, I'm not allow to do what I really want which is `impl Deserialize for Color`. If anyone knows a good way around this (I don't like the idea of introducing a NewType), please let me know.

I think at this point the only discussion left is around the `styles_for_shape` method. Do we really wanna do a color for each shape, or do some of the simplifications like in the `DefaultTheme`. I think I like things the way they are.

Also, please let me know if the tests should go somewhere else.

**Comment by thegedge** [comment]
> Unfortunately, my understanding is that since the Color type and the Deserialize trait are both external crates, I'm not allow to do what I really want which is impl Deserialize for Color. If anyone knows a good way around this (I don't like the idea of introducing a NewType), please let me know.

You're correct in your understanding, but fear not, there's a pattern to get around this: [newtype](https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html)!

File: crates/nu-cli/src/shell/palette.rs [hunk 1]
```diff
@@ -58,6 +62,281 @@ impl Palette for DefaultPalette {
     }
 }
 
+pub struct ThemedPallet {
```
Comment (thegedge) [inline crates/nu-cli/src/shell/palette.rs:65]: nit: `Palette` (pallet is something complete different 🙂)

**Comment by klnusbaum** [comment]
Alright, I've move to a new type implementation.

> One question comes to mind though. Could this actually just be a table in config? It would integrate more nicely with nushell, be one less config file someone has to think about, and also allow it to be processed in a pipeline.

I don't know much about tables in config. My plan was to have a config variable that is a file path which contains the JSON for the theme. That way, swapping a theme would be as simple as changing single config variable. Would I be able to do that with a table in config?

File: crates/nu-cli/src/shell/palette.rs [hunk 2]
```diff
@@ -89,7 +91,7 @@ impl Palette for ThemedPallet {
             FlatShape::Type => single_style_span(self.theme.r#type.bold(), shape.span),
             FlatShape::Operator => single_style_span(self.theme.operator.normal(), shape.span),
             FlatShape::DotDot => single_style_span(self.theme.dot_dot.bold(), shape.span),
-            FlatShape::Dot => single_style_span(Style::new().fg(self.theme.dot), shape.span),
+            FlatShape::Dot => single_style_span(Style::new().fg(*self.theme.dot), shape.span),
```
Comment (klnusbaum) [inline crates/nu-cli/src/shell/palette.rs:94]: Is there away to avoid having to do the explicit dereference here?

Comment (klnusbaum) [inline crates/nu-cli/src/shell/palette.rs:94]: Is there a way to avoid the explicit dereference here?

File: crates/nu-cli/src/shell/palette.rs [hunk 3]
```diff
@@ -125,6 +135,194 @@ impl Palette for ThemedPallet {
...
+        }
+    }
+}
+
 fn single_style_span(style: Style, span: Span) -> Vec<Spanned<Style>> {
     vec![Spanned::<Style> { span, item: style }]
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (klnusbaum) [inline crates/nu-cli/src/shell/palette.rs:267]: Is this where my tests should live? I poked around a little bit but I couldn't find a place to put this test that I felt good about.

Comment (jonathandturner) [reply crates/nu-cli/src/shell/palette.rs:267]: If they're unit tests, you can put them beside the code like this.

File: crates/nu-cli/src/shell/palette.rs (hunk 2 above)
Comment (jonathandturner) [reply crates/nu-cli/src/shell/palette.rs:94]: You could use `self.theme.dot.normal()` I think

**Comment by jonathandturner** [comment]
@klnusbaum - any interest in finishing this PR? We'd be happy to keep improving the theming support

**Comment by klnusbaum** [comment]
> @klnusbaum - any interest in finishing this PR? We'd be happy to keep improving the theming support

yes! Sorry this fell off my radar. I'll take a look into it today or tomorrow. I guess the last thing to do is figure out why these jobs are failing.

File: crates/nu-cli/src/shell/palette.rs [hunk 4]
```diff
@@ -69,6 +69,8 @@ pub struct ThemedPalette {
 }
 
 impl ThemedPalette {
+    // remove this once we start actually loading themes.
+    `#[allow(dead_code)]`
```
Comment (klnusbaum) [inline crates/nu-cli/src/shell/palette.rs:73]: @jonathandturner I plan to use this on my next PR. is it ok if we ignore this dead code for now? If so, I think this PR is ready to go.

**Comment by klnusbaum** [comment]
sorry about that. took me a while to fix all the lints. we should be good to go now

**Comment by jonathandturner** [comment]
Looks good!

**Pull request merged**